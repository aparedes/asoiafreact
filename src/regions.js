/* @flow */
import './regions.css'
import * as Immutable from 'immutable'
import React, { Component } from 'react'
import type { Map, Set } from 'immutable'
import { connect } from 'react-redux'

type PropsType = {
  regions: Set<string>,
  selectedRegion: string,
  setRegion: (region: string) => void,
}

class Regions extends Component {
  props: PropsType
  onClick(region) {
    this.props.setRegion(region)
  }
  render() {
    const { regions, selectedRegion } = this.props
    return (
      <div className={ 'regions' }>
        { regions.map(region => (
          <div className={ selectedRegion === region ? 'active' : '' } key={ region } onClick={ this.onClick.bind(this, region) }>
            { region.length === 0 ? 'All' : region }
          </div>
        ))}
      </div>
    )
  }
}

const mapStateToProps = (state: { houses: Map<string, *> }) => ({
  selectedRegion: state.app.region,
  regions: state.houses.getIn([ 'regions' ], new Immutable.Set()),
})
const mapDistachToProps = (dispatch: (action: Object) => void) => ({
  setRegion: (region: string) => dispatch({ type: 'SET_REGION', region }),
})
export default connect(mapStateToProps, mapDistachToProps)(Regions)
