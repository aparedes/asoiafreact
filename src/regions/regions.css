.regions {
  border: 1px solid black;
  align-items: center;
  display: flex;
  flex-direction: row;
  flex: 0 0 60px;
  min-width: 100%;
  overflow: auto;
}

.regionItem {
  flex: 0 0 150px;
  display: inline-block;
  align-self: center;
}

.regions .active {
  color: red;
}
