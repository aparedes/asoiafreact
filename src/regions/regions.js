/* @flow */
import './regions.css';
import { Set } from 'immutable';
import React, { Component, useCallback } from 'react';
import { connect } from 'react-redux';

import type { ReduxState, Dispatch } from '../store/constants/reduxTypes';

type Props = {
  regions: Set<string>,
  selectedRegion: string,
  setRegion: (region: string) => void,
};

function Region(props) {
  const { selectedRegion, region, onClick } = props;
  const onClickRegion = useCallback(() => onClick(region), [onClick, region]);
  return (
    <div
      className={selectedRegion === region ? 'regionItem active' : 'regionItem'}
      onClick={onClickRegion}
    >
      {region.length === 0 ? 'All' : region}
    </div>
  );
}
export class Regions extends Component<Props> {
  onClick = (region: string) => {
    this.props.setRegion(region);
  };
  render() {
    const { regions, selectedRegion } = this.props;
    return (
      <div className={'regions'}>
        {regions.map((region) => (
          <Region
            selectedRegion={selectedRegion}
            region={region}
            key={region}
            onClick={this.onClick}
          />
        ))}
      </div>
    );
  }
}

const mapStateToProps = (state: ReduxState) => ({
  selectedRegion: state.app.region,
  regions: state.houses.getIn(['regions'], Set()),
});
const mapDistachToProps = (dispatch: Dispatch) => ({
  setRegion: (region: string) => dispatch({ type: 'SET_REGION', region }),
});

export default connect(mapStateToProps, mapDistachToProps)(Regions);
