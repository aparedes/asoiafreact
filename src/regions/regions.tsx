/* @flow */
import './regions.css';
import React, { useCallback, useMemo } from 'react';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';

import { ReduxState } from '../store/constants/reduxTypes';
import { appActions } from '../store/reducers/app';

interface RegionProps {
  region: string;
}
function Region(props: RegionProps) {
  const { region } = props;
  const dispatch = useDispatch();
  const isSelected = useSelector<ReduxState, boolean>(
    (state) => state.app.region === region,
    shallowEqual
  );
  const onClickRegion = useCallback(
    () => dispatch(appActions.setRegion(region)),
    [dispatch, region]
  );
  return (
    <div
      data-testid="regionItem"
      // {...(/active/.test(props.className) ? { 'data-testid': 'active' } : {})}
      className={isSelected ? 'regionItem active' : 'regionItem'}
      onClick={onClickRegion}
    >
      {region.length === 0 ? 'All' : region}
    </div>
  );
}
export function Regions() {
  const regions = useSelector<ReduxState, Readonly<string[]>>(
    (state) => state.houses.regions ?? [],
    shallowEqual
  );
  const regionsMapped = useMemo(
    () =>
      regions.map((region) => {
        return <Region key={region} region={region} />;
      }),
    [regions]
  );
  return <div className={'regions'}>{regionsMapped}</div>;
}

export default Regions;
