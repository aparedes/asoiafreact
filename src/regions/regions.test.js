import { Set } from 'immutable';
import React from 'react';
import { Regions } from './regions';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

// regions: Set<string>,
// selectedRegion: string,
// setRegion: (region: string) => void,

describe('Displays the regions', () => {
  it('Region snapshot', () => {
    const component = renderer.create(
      <Regions
        regions={Set(['', 'The North', 'Dorne', 'North of The Wall'])}
        selectedRegion={''}
      />
    );
    let regions = component.toJSON();
    expect(regions).toMatchSnapshot();
  });

  //
  // regions.find('The North').onClick()
  // regions = component.toJSON()
  // expect(regions).toMatchSnapshot()
});
