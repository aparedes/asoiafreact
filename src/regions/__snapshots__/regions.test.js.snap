// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays the regions Region snapshot 1`] = `
<div
  className="regions"
>
  <div
    className="regionItem active"
    onClick={[Function]}
  >
    All
  </div>
  <div
    className="regionItem"
    onClick={[Function]}
  >
    The North
  </div>
  <div
    className="regionItem"
    onClick={[Function]}
  >
    Dorne
  </div>
  <div
    className="regionItem"
    onClick={[Function]}
  >
    North of The Wall
  </div>
</div>
`;
